webrtc/stun/stun.nim

```diff
@@ -0,0 +1,171 @@
+# Nim-WebRTC
+# Copyright (c) 2024 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+import bitops, strutils, random
+import chronos,
+       chronicles,
+       binary_serialization,
+       stew/objects,
+       stew/byteutils
+import stun_attributes
+
+export binary_serialization
+
+logScope:
+  topics = "webrtc stun"
+
+const
+  msgHeaderSize = 20
+  magicCookieSeq = @[ 0x21'u8, 0x12, 0xa4, 0x42 ]
+  magicCookie = 0x2112a442
+  BindingRequest = 0x0001'u16
+  BindingResponse = 0x0101'u16
+
+proc decode(T: typedesc[RawStunAttribute], cnt: seq[byte]): seq[RawStunAttribute] =
+  const pad = @[0, 3, 2, 1]
+  var padding = 0
+  while padding < cnt.len():
+    let attr = Binary.decode(cnt[padding ..^ 1], RawStunAttribute)
+    result.add(attr)
+    padding += 4 + attr.value.len()
+    padding += pad[padding mod 4]
+
+type
+#  Stun Header
+#  0                   1                   2                   3
+#  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+# +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+# |0 0|     STUN Message Type     |         Message Length        |
+# +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+# |                         Magic Cookie                          |
+# +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+# |                                                               |
+# |                     Transaction ID (96 bits)                  |
+# |                                                               |
+# +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+# Message type:
+#   0x0001: Binding Request
+#   0x0101: Binding Response
+#   0x0111: Binding Error Response
+#   0x0002: Shared Secret Request
+#   0x0102: Shared Secret Response
+#   0x0112: Shared Secret Error Response
+
+  RawStunMessage = object
+    msgType: uint16
+    length* {.bin_value: it.content.len().}: uint16
+    magicCookie: uint32
+    transactionId: array[12, byte] # Down from 16 to 12 bytes in RFC5389
+    content* {.bin_len: it.length.}: seq[byte]
+
+  StunMessage* = object
+    msgType*: uint16
+    transactionId*: array[12, byte]
+    attributes*: seq[RawStunAttribute]
+
+  Stun* = object
+    iceTiebreaker: uint64
+
+proc generateRandomSeq(size: int): seq[byte] =
+  result = newSeq[byte](size)
+  for i in 0..<size:
+    result[i] = rand(255).uint8
```

Comment:
Not sure how important this is, but the RFC says "As such, the transaction ID MUST be uniformly and randomly chosen from the interval 0 .. 2**96-1, and SHOULD be cryptographically random.". Is this rand cryptographically secure? Docs also says `randomize` should be called first https://nim-lang.org/docs/random.html#rand%2Cint.

Reply:
Found this https://nim-lang.org/docs/sysrand.html. @arnetheduck thoughts?

Reply:
https://github.com/vacp2p/nim-libp2p/blob/unstable/libp2p/crypto/crypto.nim#L161 - we use the cryptographically secure rng  from bearssl across the board and should do so here also.

Reply:
Done